### DISCUSSION TITLE
Correct way to combine and sign multiple inputs in a TX

### DISCUSSION THREAD

**Discussion by storopoli** [rust-bitcoin/rust-bitcoin#2732] [Q&A]
I am asking this as a last resort.
I've checked both [`bitcoin/examples/ecdsa-psbt.rs`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/examples/ecdsa-psbt.rs) and [`bitcoin/examples/taproot-psbt.rs`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/examples/taproot-psbt.rs)
along with [`bitcoin/examples/sign-tx-segwit-v0.rs`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/examples/sign-tx-segwit-v0.rs) and [`bitcoin/examples/sign-tx-taproot.rs`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/examples/sign-tx-taproot.rs).

My problem is that both examples they are signing just one input.
I need to figure it out how to sign multiple inputs (both taproot and segwit).

What I've tried so far:

1. following the `sign-tx-*` examples:
   
   1. For every input:
      - created an unsigned `Transaction` with just the single input and the outputs
      - created a `Sighash`, signed it, converted to a signed `Transaction`
      - tried to get a `Psbt` out of the signed `Transaction` (<--------- FAIL)
   2. combine all the single inputs `Psbt`s into a full-blown `Psbt`

1. following the `*-psbt` examples (in particular the `ecdsa`/segwit example):

   1. got the "wallet"'s master `Xpriv`, master `Fingerprint`
   1. for every input:
      - I got the address' `DerivationPath`, `Xpriv`, `Xpub`, `CompressedPublicKey`s `WPubkeyHash`s, `redeem_script`s
      - inserted it into the `Psbt.input` with the proper indexing:
        https://github.com/rust-bitcoin/rust-bitcoin/blob/ad212da9305a257f236746511edf707f42334361/bitcoin/examples/ecdsa-psbt.rs#L210-L221
      - signed the `Psbt`
      - finalized and emptied the `Psbt` and its `Pbst.inputs` with the proper indexing:
        https://github.com/rust-bitcoin/rust-bitcoin/blob/ad212da9305a257f236746511edf707f42334361/bitcoin/examples/ecdsa-psbt.rs#L233-L246
   1. serialized the `Psbt` (<--------- FAIL, invalid base64 in `bitcoin-cli decodepsbt`)
   
   
I would love if someone could point me towards the right direction?
I definitely intend to add a new example (and a cookbook entry) after I solve this.
It will definitely help more people in the future.

**Comment by storopoli** [comment]
I just realized that my workflow is pretty much the same as in [`bitcoin/tests/bip_174.rs`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/tests/bip_174.rs)

**Comment by tcharding** [comment]
Perhaps post the code you have that's not working. I can't work out you exact question with only links to `rust-bitcoin` code.

**Reply by storopoli** [reply]
I've managed to do it. The `PsbtExt` in miniscript simplified a lot, but I had to downgrade to Bitcoin 0.31.2 to use miniscript.
I'll implement tomorrow Taproot and if I experience no more issues I plan to make an examples PR with multiple inputs and a cookbook PR as well over the weekend 😊

**Reply by apoelstra** [reply]
Thanks! For simple cases it would be nice if it were possible to do this directly in rust-bitcoin (maybe by cribbing code from `PsbtExt`). But in general "use `miniscript`" is probably the right answer because rust-minscript has a finalizer and rust-bitcoin does not.
